src/analysis/lattices/shared.h

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_analysis_lattices_shared_h
+#define wasm_analysis_lattices_shared_h
+
+#include <cstdint>
+#include <utility>
+
+#include "../lattice.h"
+#include "bool.h"
+
+namespace wasm::analysis {
+
+// A lattice whose elements are a single ascending chain in lattice `L`.
+// Internally, there is only ever a single monotonically increasing element of L
+// materialized. Dereferencing any element of the Shared lattice will produce
+// the current value of that single element of L, which is generally safe
+// because the current value always overapproximates the value at the time of
+// the Shared element's construction.
+template<Lattice L> struct Shared {
+  class Element {
+    const typename L::Element* val;
+    uint32_t seq = 0;
```

Comment:
```suggestion
    Index seq = 0;
```
Index seems appropriate as this number is tied to the number of basic blocks (and iterations on them), which is tied to maximum sizes in wasm files. That is, if wasm supports more then 32 bits here, we'd want this type to change, which `Index` will do for us.

Reply:
This number is not tied to the number of basic blocks. It is the number of times the shared value is increased, which can be dramatically higher than the number of basic blocks. Since analysis continues to a fixed point, it can visit basic blocks an unbounded number of times and each visit to a basic block may potentially increase the value multiple times.

Since this isn't actually an index into anything that is ever materialized, I would prefer to keep it a uint32_t for now, or possibly make it a uint64_t right now.

Reply:
sgtm. How about using a name for it for convenience later if we need to change it, `using Seq = uint32_t;` ?